src/http/client.cc

```diff
@@ -312,6 +312,15 @@ auto client::with_connection(Fn&& fn) {
     });
 }
 
+struct transport_error : public std::exception {
+    unsigned requests_submitted;
```

Comment:
The catcher of this exception checks how many requests were submitted via the connection it tried to use. If it was an connection with more than one request, it means that the connection was picked up from cache and it make sense to replay the request via fresh new one. If the connection tried to serve just one request, it means that it was _not_ a connection from pool, but a new connection. In that case it doesn't make sense to retry request via another new connection.

Reply:
> I still don't understand both points:
> 
>     1. If the connection was from the pool, why do we need "a new connection" and can't take another connection from the pool?

That's what #1847 does -- it picks another connection from whatever it is -- from the pool or makes a new one. That approach also works, but needs configuration from caller -- how many times make retry attempts. This, in turn, is because retrying once over connection from poll doesn't really work -- retrying from the connection from poll fails again quite often and needs 2nd and sometimes 3rd retry.

This PR retries once, because in my experiments retrying over fresh connection works in 99.99% of the cases.

>     2. If this was the first request on this connection, how does it's prove it's not from the pool? Can't it be the first request made on a connection that is / will be in the pool?

The connection usage logic is like this:

```
    con = pop_from_pool()
    if (con == nullptr)
        con = make_new_con()
    con.requests_submitted++
    con.make_request()
    push_to_pool(con)
```

in case con.make_request() throws and con.request == 1 it means that the connection came from make_new_con() and if con.request == 2 it means that the connection came from pool

Reply:
>  ... and if con.request == 2 it means that the connection came from pool

fix: ... and if con.request >= 2 ...

Reply:
I still don't understand this. After you pop_from_pool(), the connection - no matter where it came from - is no longer in the pool. Why is it important to you to remember if it came from the pool or didn't? How is this knowledge relevant for deciding what to do on retry? On retry you need to close the old connection (it's broken...) and create a new one. Unconditionally... Who cares whether the original connection came from the pool or not?

Reply:
This logic is removed, so resolving the conversation